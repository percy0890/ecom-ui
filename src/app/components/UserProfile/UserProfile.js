import React from 'react';
import PropTypes from 'prop-types';
import './UserProfile.scss';
import Button from '../Button';

export class UserProfile extends React.PureComponent {
  state = {
    isBtnActive: false,
  };

  toggleBtn = () => {
    this.setState(
      prevState => ({
        isBtnActive: !prevState.isBtnActive,
      }),
      () => {
        if (this.state.isBtnActive) {
          document.addEventListener(
            'mousedown',
            this.handleOutsideClick,
            false,
          );
        } else {
          document.removeEventListener(
            'mousedown',
            this.handleOutsideClick,
            false,
          );
        }
      },
    );
  };

  handleOutsideClick = e => {
    if (this.profileDDNode && !this.profileDDNode.contains(e.target)) {
      this.setState({
        isBtnActive: false,
      });
    }
  };

  onRefreshMaster = () => {
    this.setState({
      isBtnActive: false,
    });
    this.props.onRefreshMaster();
  };

  onLogout = () => {
    this.setState({
      isBtnActive: false,
    });
    this.props.onLogout();
  };

  render() {
    const { customClass, name, profileImage } = this.props;
    return (
      <div className="profile-container">
        <div
          className={`profile-btn-container ${
            this.state.isBtnActive ? 'active' : ''
          } ${customClass}`}
          onClick={this.toggleBtn}
          role="presentation"
        >
          <span className="profile-name">{name}</span>
          <Button
            buttonType="transparent"
            customClass="profile-button"
            icon={profileImage}
          />
        </div>
        {this.state.isBtnActive && (
          <div
            className="btn-dd-container"
            ref={elem => {
              this.profileDDNode = elem;
            }}
          >
            <div
              className="dd-item"
              onClick={this.onLogout}
              role="presentation"
            >
              Logout
            </div>
          </div>
        )}
      </div>
    );
  }
}

UserProfile.propTypes = {
  customClass: PropTypes.string,
  profileImage: PropTypes.string,
  name: PropTypes.string,
  onRefreshMaster: PropTypes.func,
  onLogout: PropTypes.func,
};

export default UserProfile;
