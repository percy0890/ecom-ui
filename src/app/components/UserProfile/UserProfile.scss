@import '~app/styles/core.scss';

.profile-container {
  position: relative;
  cursor: pointer;

  .profile-btn-container {
    display: flex;
    align-items: center;
    border: 1px solid transparent;
    padding: 0 1.5rem;
    border-radius: 4px;

    &.active {
      border: 1px solid #ccc;
      box-shadow: $app-shadow-element;
    }

    .profile-name {
      display: flex;
      align-items: center;
    }

    .profile-button {
      background-color: transparent;
      border: none;
      padding: 0;
      margin-left: 1rem;
      height: 3.8rem;

      @include media(desktop down) {
        margin-left: 1rem;
      }

      .icon {
        height: 2rem;
      }
    }
  }

  .btn-dd-container {
    position: absolute;
    right: 0;
    top: 4.5rem;
    min-width: 110px;
    box-shadow: $app-shadow-container;
    border-radius: 4px;
    z-index: 10;
    background: $app-bg-white;
    border: 1px solid #ddd;

    .dd-item {
      padding: 1rem 2rem;
      color: $app-text-primary;
      border-bottom: 1px solid #ddd;

      &.static {
        cursor: default;
      }

      &:first-child {
        border-radius: 4px 4px 0 0;
      }

      &:last-child {
        border-radius: 0 0 4px 4px;
        border-bottom: 0;

        &:first-child {
          border-radius: 4px;
        }
      }

      &:not(.static):hover {
        background: $brand-color-primary;
        color: $app-text-white;
      }
    }
  }
}
