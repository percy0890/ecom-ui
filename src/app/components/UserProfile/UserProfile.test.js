import React from 'react';
import { shallow } from 'enzyme';
import UserProfile from './UserProfile';

describe('<UserProfile />', () => {
  let component;

  beforeEach(() => {
    component = shallow(
      <UserProfile onRefreshMaster={jest.fn()} onLogout={jest.fn()} />,
    );
  });

  it('renders correctly', () => {
    expect(component).toMatchSnapshot();
  });

  it('should render user dropdown on clicking name/pic', () => {
    component
      .find('.profile-btn-container')
      .at(0)
      .simulate('click');
    expect(component.state().isBtnActive).toBe(true);
  });

  it('should hide user dropdown when isBtnActive is set to false', () => {
    component.setState({
      isBtnActive: true,
    });
    component
      .find('.profile-btn-container')
      .at(0)
      .simulate('click');
    expect(component.state().isBtnActive).toBe(false);
  });

  it('should call f(onLogout) on clicking `Logout` from user dropdown menu', () => {
    component.setState({
      isBtnActive: true,
    });
    const onLogout = jest.spyOn(component.instance().props, 'onLogout');
    component
      .find('.dd-item')
      .at(0)
      .simulate('click');
    expect(onLogout).toHaveBeenCalled();
  });

  it('Set state  on accordion to display details', () => {
    component.setState({ isBtnActive: true });
    const profileDDNode = component.find('.btn-dd-container');
    component.instance().profileDDNode = profileDDNode;
    component.instance().handleOutsideClick({ target: <div> </div> });
    expect(component.state().isBtnActive).toBe(false);
  });
  it('Set state  on accordion to display without profilenode', () => {
    component.setState({ isBtnActive: true });
    component.instance().profileDDNode = null;
    component.instance().handleOutsideClick({ target: <div> </div> });
    expect(component.state().isBtnActive).toBe(true);
  });

  it('Set state  on onRefreshMaster ', () => {
    component.setState({ isBtnActive: true });
    component.instance().onRefreshMaster();
    expect(component.state().isBtnActive).toBe(false);
  });
});
