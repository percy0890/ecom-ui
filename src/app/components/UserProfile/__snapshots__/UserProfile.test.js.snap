// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UserProfile /> renders correctly 1`] = `
<div
  className="profile-container"
>
  <div
    className="profile-btn-container  undefined"
    onClick={[Function]}
    role="presentation"
  >
    <span
      className="profile-name"
    />
    <Button
      buttonType="transparent"
      customClass="profile-button"
    />
  </div>
</div>
`;
